pkg/upgrade/upgrade_state.go

```diff
@@ -0,0 +1,328 @@
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to uncordon nodes")
+		return err
+	}
+	m.Log.V(consts.LogLevelInfo).Info("State Manager, finished processing")
+	return nil
+}
+
+// ProcessDoneOrUnknownNodes iterates over UpgradeStateDone or UpgradeStateUnknown nodes and determines
+// whether each specific node should be in UpgradeStateUpgradeRequired or UpgradeStateDone state.
+func (m *ClusterUpgradeStateManager) ProcessDoneOrUnknownNodes(
+	ctx context.Context, currentClusterState *ClusterUpgradeState, nodeStateName string) error {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessDoneOrUnknownNodes")
+
+	for _, nodeState := range currentClusterState.NodeStates[nodeStateName] {
+		if GetPodGeneration(nodeState.DriverPod, m.Log) != nodeState.DriverDaemonSet.GetGeneration() {
+			err := m.NodeUpgradeStateProvider.ChangeNodeUpgradeState(ctx, nodeState.Node, UpgradeStateUpgradeRequired)
+			if err != nil {
+				m.Log.V(consts.LogLevelError).Error(
+					err, "Failed to change node upgrade state", "state", UpgradeStateUpgradeRequired)
+				return err
+			}
+		} else if nodeStateName == UpgradeStateUnknown {
+			err := m.NodeUpgradeStateProvider.ChangeNodeUpgradeState(ctx, nodeState.Node, UpgradeStateDone)
+			if err != nil {
+				m.Log.V(consts.LogLevelError).Error(
+					err, "Failed to change node upgrade state", "state", UpgradeStateDone)
+				return err
+			}
+		}
+		m.Log.V(consts.LogLevelInfo).Info("Node in done state, upgrade not required",
+			"node", nodeState.Node.Name)
+	}
+	return nil
+}
+
+// ProcessUpgradeRequiredNodes processes UpgradeStateUpgradeRequired nodes and moves them to UpgradeStateDrain until
+// the limit on max parallel upgrades is reached.
+func (m *ClusterUpgradeStateManager) ProcessUpgradeRequiredNodes(
+	ctx context.Context, currentClusterState *ClusterUpgradeState, limit int) error {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessUpgradeRequiredNodes")
+	for _, nodeState := range currentClusterState.NodeStates[UpgradeStateUpgradeRequired] {
+		if limit <= 0 {
+			m.Log.V(consts.LogLevelInfo).Info("Limit for new upgrades is exceeded, skipping the iteration")
+			break
+		}
+
+		err := m.NodeUpgradeStateProvider.ChangeNodeUpgradeState(ctx, nodeState.Node, UpgradeStateDrain)
+		if err == nil {
+			limit--
+			m.Log.V(consts.LogLevelInfo).Info("Node waiting for drain",
+				"node", nodeState.Node.Name)
+		} else {
+			m.Log.V(consts.LogLevelError).Error(
+				err, "Failed to change node upgrade state", "state", UpgradeStateDrain)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ProcessDrainNodes schedules UpgradeStateDrain nodes for drain.
+// If drain is disabled by upgrade policy, moves the nodes straight to UpgradeStatePodRestart state.
+func (m *ClusterUpgradeStateManager) ProcessDrainNodes(
+	ctx context.Context, currentClusterState *ClusterUpgradeState, drainSpec *v1alpha1.DrainSpec) error {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessDrainNodes")
+	if !drainSpec.Enable {
```

Comment:
I think drainSpec can be nil here? Is it?

Reply:
Added check